File: java/Sortings/Merge.java
package Sortings;

import java.util.Arrays;
import java.util.Comparator;

public class Merge {
    public static <T> void sort(T[] arr, Comparator<T> comp) {
        var result = sortHelper(arr, comp);
        System.arraycopy(result, 0, arr, 0, result.length);
    }

    private static <T> T[] sortHelper(T[] arr, Comparator<T> comp) {
        if (arr.length == 1) return arr;
        int half = arr.length / 2;
        var first = sortHelper(Arrays.copyOfRange(arr, 0, half), comp);
        var second = sortHelper(Arrays.copyOfRange(arr, half, arr.length), comp);
        return merge(first, second, comp);
    }


    @SuppressWarnings("unchecked")
    private static <T> T[] merge(T[] arr1, T[] arr2, Comparator<T> comp) {
        T[] arr = (T[]) new Object[arr1.length + arr2.length];
        int i = 0, j = 0, k = 0;
        while (i < arr1.length && j < arr2.length) {
            if (comp.compare(arr1[i], arr2[j]) < 0) arr[k++] = arr1[i++];
            else arr[k++] = arr2[j++];
        }
        while (i < arr1.length) {
            arr[k++] = arr1[i++];
        }
        while (j < arr2.length) {
            arr[k++] = arr2[j++];
        }
        return arr;
    }

    public static <T extends Comparable<T>> void sort(T[] arr) {
        sort(arr, Comparable::compareTo);
    }
}

File: typescript/src/sortings/merge.ts
import { CompareFn } from "../util.ts";

function sortHelper<T>(
  arr: T[],
  low: number,
  high: number,
  compare: CompareFn<T>,
) {
  if (low >= high) return;
  const half = Math.round((low + high) / 2);
  sortHelper(arr, low, half - 1, compare);
  sortHelper(arr, half, high, compare);
  merge(
    { arr, start: low, end: half - 1 },
    { arr, start: half, end: high },
    compare,
  );
}

type ArrayWithRange<T> = {
  arr: T[];
  start: number;
  end: number;
};

function merge<T>(
  a: ArrayWithRange<T>,
  b: ArrayWithRange<T>,
  compare: CompareFn<T>,
) {
  const ansArr: T[] = [];
  let i = a.start, j = b.start;
  while (i <= a.end && j <= b.end) {
    if (compare(a.arr[i], b.arr[j]) <= 0) {
      ansArr.push(a.arr[i++]);
    } else {
      ansArr.push(b.arr[j++]);
    }
  }
  while (i <= a.end) ansArr.push(a.arr[i++]);
  while (j <= b.end) ansArr.push(b.arr[j++]);
  for (let k = 0; k < ansArr.length; k++) {
    a.arr[a.start + k] = ansArr[k];
  }
  return ansArr;
}

export default function sort<T>(arr: T[], compare: CompareFn<T>) {
  sortHelper(arr, 0, arr.length - 1, compare);
}
